File: python/src/reachy_jev/questions.py
"""Versioned question banks and the TypeSafe request-wire projection."""

from __future__ import annotations

import re
from typing import Any

_PERSON = re.compile(r"p[1-9]\Z")
_KEY = re.compile(r"[a-z][a-z0-9_]*\Z")
_BANK = re.compile(r"[a-z][a-z0-9_.-]*\Z")
_VERSION = re.compile(r"\d+\.\d+\.\d+\Z")


def _nonempty(value: Any) -> bool:
    return isinstance(value, str) and bool(value.strip())


def expand_bank(bank: dict[str, Any], person_ids: list[str]) -> dict[str, dict[str, Any]]:
    if (
        not isinstance(bank, dict)
        or not isinstance(bank.get("bank"), str)
        or not _BANK.fullmatch(bank["bank"])
    ):
        raise ValueError("invalid bank name")
    if not isinstance(bank.get("version"), str) or not _VERSION.fullmatch(bank["version"]):
        raise ValueError("invalid bank version")
    questions = bank.get("questions")
    if not isinstance(questions, dict) or not questions:
        raise ValueError("empty question bank")
    if not isinstance(person_ids, list):
        raise ValueError("invalid person IDs")
    ids = list(dict.fromkeys(person_ids))
    if any(not isinstance(person_id, str) or not _PERSON.fullmatch(person_id) for person_id in ids):
        raise ValueError("invalid person ID")
    expanded = {}
    for key, question in questions.items():
        if not isinstance(key, str) or not _KEY.fullmatch(key) or not isinstance(question, dict):
            raise ValueError("invalid question")
        instructions = question.get("instructions")
        if not _nonempty(instructions):
            raise ValueError("invalid question instructions")
        kind = question.get("type")
        if kind == "choice":
            source = question.get("options")
            if set(question) - {"type", "instructions", "options"} or not isinstance(source, list):
                raise ValueError("invalid choice options")
            options = [
                person_id for option in source for person_id in (ids if option == "$people.ids" else [option])
            ]
            if (
                not options
                or any(not _nonempty(option) for option in options)
                or len(set(options)) != len(options)
            ):
                raise ValueError(f"invalid choice options: {key}")
            expanded[key] = {**question, "options": options}
        elif kind == "score":
            levels = question.get("levels")
            if (
                set(question) - {"type", "instructions", "levels"}
                or not isinstance(levels, list)
                or len(levels) < 2
                or any(not _nonempty(level) for level in levels)
                or len(set(levels)) != len(levels)
            ):
                raise ValueError(f"invalid score levels: {key}")
            expanded[key] = dict(question)
        elif kind == "noul":
            if set(question) - {"type", "instructions", "criteria"} or (
                "criteria" in question and not _nonempty(question["criteria"])
            ):
                raise ValueError(f"invalid noul criteria: {key}")
            expanded[key] = dict(question)
        else:
            raise ValueError(f"unknown question type: {key}")
    return expanded


def to_typesafe_questions(bank: dict[str, Any], person_ids: list[str]) -> dict[str, dict[str, Any]]:
    """Keep input text in state data; only fixed bank content becomes instructions."""
    wire = {}
    for key, question in expand_bank(bank, person_ids).items():
        instructions = question["instructions"]
        if question["type"] == "noul" and question.get("criteria"):
            instructions += f" Criteria: {question['criteria']}"
        if question["type"] == "choice":
            wire[key] = {
                "type": "choice",
                "instructions": instructions,
                "criteria": dict.fromkeys(question["options"]),
            }
        elif question["type"] == "score":
            wire[key] = {"type": "score", "instructions": instructions, "criteria": list(question["levels"])}
        else:
            wire[key] = {"type": "noul", "instructions": instructions}
    return wire

File: src/questions.ts
export type NoulQuestion = { type: "noul"; instructions: string; criteria?: string };
export type ChoiceQuestion = { type: "choice"; instructions: string; options: readonly string[] };
export type ScoreQuestion = { type: "score"; instructions: string; levels: readonly string[] };
export type Question = NoulQuestion | ChoiceQuestion | ScoreQuestion;
export interface QuestionBank { bank: string; version: string; questions: Record<string, Question> }
export type WireQuestion =
  | { type: "noul"; instructions: string }
  | { type: "choice"; instructions: string; criteria: Record<string, null> }
  | { type: "score"; instructions: string; criteria: readonly [string, string, ...string[]] };

function record(value: unknown): value is Record<string, unknown> {
  return value !== null && typeof value === "object" && !Array.isArray(value);
}
function nonempty(value: unknown): value is string {
  return typeof value === "string" && value.trim().length > 0;
}
function fields(value: Record<string, unknown>, allowed: readonly string[]): boolean {
  return Object.keys(value).every((key) => allowed.includes(key));
}

/** Expand the only dynamic token supported by the bank format. Never interpolate transcript text. */
export function expandBank(bank: QuestionBank, personIds: readonly string[]): Record<string, Question> {
  if (!record(bank) || typeof bank.bank !== "string" || !/^[a-z][a-z0-9_.-]*$/.test(bank.bank) || typeof bank.version !== "string" || !/^\d+\.\d+\.\d+$/.test(bank.version)) throw new TypeError("invalid bank identity");
  if (!record(bank.questions) || !Object.keys(bank.questions).length) throw new TypeError("empty question bank");
  if (!Array.isArray(personIds)) throw new TypeError("invalid person IDs");
  const ids = [...new Set(personIds)];
  if (ids.some((id) => typeof id !== "string" || !/^p[1-9]$/.test(id))) throw new TypeError("invalid person ID");
  return Object.fromEntries(Object.entries(bank.questions).map(([key, question]) => {
    if (!/^[a-z][a-z0-9_]*$/.test(key) || !record(question) || !nonempty(question.instructions)) throw new TypeError(`invalid question: ${key}`);
    if (question.type === "choice") {
      if (!fields(question, ["type", "instructions", "options"]) || !Array.isArray(question.options) || question.options.some((option) => !nonempty(option))) throw new TypeError(`invalid choice options: ${key}`);
      const options = question.options.flatMap((option) => option === "$people.ids" ? ids : [option]);
      if (!options.length || new Set(options).size !== options.length) throw new TypeError(`invalid choice options: ${key}`);
      return [key, { ...question, options }];
    }
    if (question.type === "score") {
      if (!fields(question, ["type", "instructions", "levels"]) || !Array.isArray(question.levels) || question.levels.length < 2 || question.levels.some((level) => !nonempty(level)) || new Set(question.levels).size !== question.levels.length) throw new TypeError(`invalid score levels: ${key}`);
      return [key, question];
    }
    if (question.type === "noul") {
      if (!fields(question, ["type", "instructions", "criteria"]) || (question.criteria !== undefined && !nonempty(question.criteria))) throw new TypeError(`invalid noul criteria: ${key}`);
      return [key, question];
    }
    throw new TypeError(`unknown question type: ${key}`);
  }));
}

/** Convert the human-readable bank into the TypeSafe JS SDK's actual request shape. */
export function toTypeSafeQuestions(bank: QuestionBank, personIds: readonly string[]): Record<string, WireQuestion> {
  const expanded = expandBank(bank, personIds);
  const wire: Record<string, WireQuestion> = {};
  for (const [key, question] of Object.entries(expanded)) {
    const instructions = question.type === "noul" && question.criteria
      ? `${question.instructions} Criteria: ${question.criteria}`
      : question.instructions;
    if (question.type === "choice") {
      wire[key] = { type: "choice", instructions, criteria: Object.fromEntries(question.options.map((option) => [option, null])) };
    } else if (question.type === "score") {
      // expandBank already requires at least two levels; retain that invariant in the type.
      wire[key] = { type: "score", instructions, criteria: [...question.levels] as [string, string, ...string[]] };
    } else {
      wire[key] = { type: "noul", instructions };
    }
  }
  return wire;
}
